File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/RuleSelectionPolicy.java
package com.foundgine.core.semantic.planning;

import java.util.Objects;

/** Policy controlling deterministic selection among simultaneously applicable rewrite rules. */
public record RuleSelectionPolicy(boolean preferHigherBenefit,boolean penalizeRewriteCost,double minimumScore) {
    public RuleSelectionPolicy(){this(true,true,Double.NEGATIVE_INFINITY);}
    public RuleSelectionPolicy validate(){if(Double.isNaN(minimumScore)||Double.isInfinite(minimumScore)&&minimumScore>0)throw new IllegalArgumentException("Minimum score is invalid.");return this;}
    public RewriteScore score(IPlanRewriteRule rule){Objects.requireNonNull(rule);var benefit=RewriteBenefit.from(Math.max(0d,rule.benefitEstimate()));var cost=RewriteCost.from(Math.max(0d,rule.costImpact()));if(!penalizeRewriteCost)return new RewriteScore(preferHigherBenefit?benefit.estimatedBenefit():-benefit.estimatedBenefit());var score=RewriteScore.calculate(benefit,cost).value();return new RewriteScore(preferHigherBenefit?score:-score);}
}

File: src/csharp/Foundgine.Core/Semantic/Planning/RuleSelectionPolicy.cs
namespace Foundgine.Core.Semantic.Planning;

/// <summary>Policy controlling deterministic selection among simultaneously applicable rewrite rules.</summary>
public sealed record RuleSelectionPolicy(
    bool PreferHigherBenefit = true,
    bool PenalizeRewriteCost = true,
    double MinimumScore = double.NegativeInfinity)
{
    public RuleSelectionPolicy Validate()
    {
        if (double.IsNaN(MinimumScore) || double.IsPositiveInfinity(MinimumScore))
            throw new ArgumentOutOfRangeException(nameof(MinimumScore));
        return this;
    }

    public RewriteScore Score(IPlanRewriteRule rule)
    {
        ArgumentNullException.ThrowIfNull(rule);
        var benefit = RewriteBenefit.From(Math.Max(0d, rule.BenefitEstimate));
        var cost = RewriteCost.From(Math.Max(0d, rule.CostImpact));
        if (!PenalizeRewriteCost)
            return new RewriteScore(PreferHigherBenefit ? benefit.EstimatedBenefit : -benefit.EstimatedBenefit);
        var score = RewriteScore.Calculate(benefit, cost).Value;
        return new RewriteScore(PreferHigherBenefit ? score : -score);
    }
}
